File: libs/common/estategap_common/models/zone.py
"""Zone Pydantic model."""

from __future__ import annotations

from decimal import Decimal
from enum import IntEnum
from uuid import UUID

from pydantic import field_validator

from ._base import AwareDatetime, EstateGapModel, validate_country_code


class ZoneLevel(IntEnum):
    COUNTRY = 0
    REGION = 1
    PROVINCE = 2
    CITY = 3
    NEIGHBOURHOOD = 4


class Zone(EstateGapModel):
    id: UUID
    name: str
    name_local: str | None = None
    country_code: str
    level: ZoneLevel
    parent_id: UUID | None = None
    geometry_wkt: str | None = None
    bbox_wkt: str | None = None
    population: int | None = None
    area_km2: Decimal | None = None
    slug: str | None = None
    osm_id: int | None = None
    created_at: AwareDatetime | None = None
    updated_at: AwareDatetime | None = None

    @field_validator("country_code")
    @classmethod
    def _validate_country_code(cls, value: str) -> str:
        return validate_country_code(value)


__all__ = ["Zone", "ZoneLevel"]

File: libs/pkg/models/zone.go
package models

import (
	"github.com/jackc/pgx/v5/pgtype"
	"github.com/shopspring/decimal"
)

type Zone struct {
	ID          pgtype.UUID         `json:"id" db:"id"`
	Name        string              `json:"name" db:"name"`
	NameLocal   *string             `json:"name_local" db:"name_local"`
	CountryCode string              `json:"country_code" db:"country_code"`
	Level       int16               `json:"level" db:"level"`
	ParentID    *pgtype.UUID        `json:"parent_id" db:"parent_id"`
	GeometryWKT *string             `json:"geometry_wkt" db:"geometry_wkt"`
	BBoxWKT     *string             `json:"bbox_wkt" db:"bbox_wkt"`
	Population  *int32              `json:"population" db:"population"`
	AreaKm2     *decimal.Decimal    `json:"area_km2" db:"area_km2"`
	Slug        *string             `json:"slug" db:"slug"`
	OsmID       *int64              `json:"osm_id" db:"osm_id"`
	CreatedAt   *pgtype.Timestamptz `json:"created_at" db:"created_at"`
	UpdatedAt   *pgtype.Timestamptz `json:"updated_at" db:"updated_at"`
}
